Added `Joint_Handmade` and `MarginalDistribution` in stats

akash9712 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->


#### Brief description of what is fixed or changed
Added `JointRV`, which allows the user to create a joint random variable
with each of its component varying over the real line. Also added
`JointDistributionHandmade` for this purpose. 
Added `MarginalDistribution` to allow random variables with components marginalized out. This also adds the ability to handle compound random variables by marginalizing over latent distributions, with or without a given condition.

#### Other comments
Same code as #14951 is added, which has to be closed due to some extra commits showing up.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. If there is no release notes entry for this PR,
write "NO ENTRY". The bot will check your release notes automatically to see
if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
    * Add `JointRV` to allow user to create Joint random variables
    * Add `MarginalDistribution`, `CompoundDistribution`
    * Ability to handle compound distributions 
<!-- END RELEASE NOTES -->

akash9712 committed:
Added `JointRV`

Added `JointRV`, which allows the user to create a joint random variable
with each of its component varying over the real line. Added
`JointDistributionHandmade` for this purpose. Also made minor
modification in core/function.py to allow `Lambda` to accept indexed
symbols. Added tests.

akash9712 committed:
Fix in `JointRV`

Fix in `JointRV` to order indexed symbols extracted by their index,
instead of using the set as it is. Also replaced symbols by indexed
symbols in `marginal_distribution`. `Indexed` are not used directly and
instead replace `Symbol` towards the end of the function because
`integrate` raised error in some cases when the variable of integration
was of type `Indexed`.

akash9712 committed:
[WIP]Added `MarginalDistribution`

Added class `MarginalDistribution` and the ability to handle compound
distributions using this.

akash9712 committed:
Modifications in tests in stats

Modified some tests, and added `evaluate` flag for marginal
distributions.

akash9712 committed:
Added test for density of compound distribution

akash9712 committed:
Partially fixed failing tests in stats.

Fixed some of the previously failing tests in stats module, by
constructing marginal distribution with existing distributions
instead of PDFs.

akash9712 committed:
Fixed `domain` for `JointPSpace`

Fixed `domain` for JointPSpace and made some corrections to `pdf` in
`MarginalDistribution`.

akash9712 committed:
Minor fixes

Removed trailing whitespaces, corrected import statements in stats.

Upabjojr commented:
When I look at
```python
In [11]:     x = Symbol('x')
    ...:     l = Symbol('l', positive=True)
    ...:     rate = Beta(l, 2, 3)
    ...:     X = Poisson(x, rate)

In [12]: X
Out[12]: x

In [13]: X.pspace
Out[13]: JointPSpace(x, MarginalDistribution(PoissonDistribution(l), (x,)))
```
I mean, `PoissonDistribution(l)` where `l` is a random symbol is a bit weird, because the distribution is not a single continuous distribution as the class structure would lead to think:
```python
In [23]: pd
Out[23]: PoissonDistribution(l)

In [24]: type(pd)
Out[24]: sympy.stats.drv_types.PoissonDistribution

In [25]: type(pd).__mro__
Out[25]: 
(sympy.stats.drv_types.PoissonDistribution,
 sympy.stats.drv.SingleDiscreteDistribution,
 sympy.core.basic.Basic,
 sympy.stats.rv.NamedArgsMixin,
 object)
```

Upabjojr commented:
Can we represent the joint distribution of a compound random variable?

Upabjojr commented:
I guess the best way to represent a compound distribution is by having a `CompoundDistribution` object.

akash9712 committed:
Added `CompoundDistribution`

Added class `CompoundDistribution`, which requires a distribution with a
random parameter for construction.

akash9712 committed:
Modifications for compound random variables

In case the user creates a compound random variable, the distribution is
of the type `CompoundDistribution`. Added and modified some tests.

ArifAhmed1995 commented on sympy/stats/joint_rv.py:
```diff
@@ -46,26 +57,49 @@ def value(self):
     def component_count(self):
-        return len(self.distribution.set.args)
+        _set = self.distribution.set
+        if isinstance(_set, ProductSet):
```
You could use ternary operator here.
`return len(_set.args) if isinstance(_set, ProductSet) else 1`
Doubt would affect execution time though.

ArifAhmed1995 commented on sympy/stats/joint_rv.py:
```diff
@@ -162,3 +207,115 @@ def marginal_distribution(rv, *indices):
+    distributed according to some given distribution.
+    """
+    def __new__(cls, dist):
+        assert isinstance(dist, (ContinuousDistribution, DiscreteDistribution))
```
Might be beneficial for end-user to get a a better explicit error message if instantiation is attempted to be done from an illegal distribution.
Such as:
```
if not isinstance(dist, (ContinuousDistribution, DiscreteDistribution)):
    raise ValueError('CompoundDistribution can only be initialized from ContinuousDistribution or DiscreteDistribution')
```
or something similar.

ArifAhmed1995 commented on sympy/stats/joint_rv.py:
```diff
@@ -162,3 +207,115 @@ def marginal_distribution(rv, *indices):
+class MarginalDistribution(Basic):
+
+    def __new__(cls,dist, rvs):
+        assert all([isinstance(rv, (Indexed, RandomSymbol))] for rv in rvs)
```
Same as discussed earlier. Better for end-user to get a proper error message.

akash9712 committed:
Added error messages in stats/joint_rv.py

Added error messages and fixed one incorrect test in
test_joint_rv.py

akash9712 commented on sympy/stats/tests/test_joint_rv.py:
```diff
@@ -58,3 +59,12 @@ def test_JointPSpace_margial_distribution():
+    x1, x2 = (Indexed('x', i) for i in (1, 2))
+    pdf = exp(-x1**2/2 + x1 - x2**2/2 - S(1)/2)/(2*pi)
+    X = JointRV('x', pdf)
+    density(X)(1, 2) == exp(-1)/(2*pi)
```
Thanks, that was an incorrect test. Fixed it now.

Upabjojr commented on sympy/stats/tests/test_mix.py:
```diff
@@ -1,15 +1,28 @@
+
+def test_compound_distribution():
+    Y = Poisson('Y', 1)
+    Z = Poisson('Z', Y)
```
I tried `Z.pspace.distribution.pdf(x)` but I get an exception.
